File: leetcode/Search_in_Rotated_Sorted_Array.java
import java.util.*;

public class Solution {
    int binarySearch(int[] A, int target, int a, int b) {
        while (a < b) {
            int m = (a + b) / 2;
            if (A[m] == target) return m;
            else if (A[m] > target) b = m;
            else a = m + 1;
        }
        return -1;
    }
    public int search(int[] A, int target) {
        if (A[0] > A[A.length - 1]) {
            int sec = A.length - 1;
            int a = 0;
            int b = A.length - 1;
            while (a + 1 < b) {
                int m = (a + b) / 2;
                if (A[m] > A[0]) a = m;
                if (A[m] < A[A.length - 1]) b = m;
            }
            int res = binarySearch(A, target, 0, b);
            if (res != -1) return res;
            return binarySearch(A, target, b, A.length);
        } else {
            return binarySearch(A, target, 0, A.length);
        }
    }
}

File: leetcode/Search_in_Rotated_Sorted_Array.cpp
class Solution {
    int binarySearch(int A[], int first, int last, int target) {
        if (first == last) {
            if (A[first] == target)
                return first;
            return -1;
        }
        if (first + 1 == last) {
            if (A[first] == target)
                return first;
            return binarySearch(A, last, last, target);
        }
        int middle = (first + last) / 2;
        if (A[middle] == target)
            return middle;
        // Only one half can be regular
        if (A[first] <= A[middle - 1])  // First half is regular
            if (A[first] <= target && A[middle - 1] >= target)
                return binarySearch(A, first, middle - 1, target);
            else
                return binarySearch(A, middle + 1, last, target);
        else  // Second half is regular
            if (A[middle + 1] <= target && A[last] >= target)
                return binarySearch(A, middle + 1, last, target);
            else
                return binarySearch(A, first, middle - 1, target);
    }
public:
    int search(int A[], int n, int target) {
        if (n < 1)
            return -1;
        return binarySearch(A, 0, n - 1, target);
    }
};
